File: BOJ/07. DP/2294.py
# 동전 2

import sys
input = sys.stdin.readline

n, k = map(int, input().split(' '))
values = [int(input()) for _ in range(n)]

dp = [1e9 for _ in range(k+1)]

for value in values:
    if value <= k:
        dp[value] = 1

for i in range(1, k+1):
    for value in values:
         if i - value > 0:
            dp[i] = min(dp[i-value] + 1, dp[i])

if dp[k] == 1e9:
    print(-1)
else:
    print(dp[k])

File: Javascript/BOJ/2294.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(filePath).toString().trim().split("\n");
const [n, k] = input.shift().split(" ").map(Number);
const coin = input.map(Number);

const dp = Array.from({ length: k + 1 }, () => Infinity);

for (let c of coin) {
    dp[c] = 1;
}

for (let i = 1; i <= k; i++) {
    for (let j = 0; j < n; j++) {
        if (i - coin[j] < 0) continue;

        dp[i] = Math.min(dp[i], dp[i - coin[j]] + 1);
    }
}

console.log(dp[k] === Infinity ? -1 : dp[k]);
